src/lib/y2network/autoinst_profile/section_with_attributes.rb

```diff
@@ -0,0 +1,189 @@
+# encoding: utf-8
+
+# Copyright (c) [2019] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+
+module Y2Network
+  module AutoinstProfile
+    # Abstract base class for some AutoYaST profile sections
+    #
+    # @todo This class is supposed to live in autoyast2 or yast2. When done,
+    #   please, adapt yast2-storage-ng accordingly.
+    class SectionWithAttributes
+      include Yast::Logger
+
+      class << self
+        # Description of the attributes in the section.
+        #
+        # To be defined by each subclass. Each entry contains a hash with the
+        # mandatory key :name and an optional key :xml_name.
+        #
+        # @return [Array<Hash>]
+        def attributes
+          []
+        end
+
+        # Creates an instance based on the profile representation used by the
+        # AutoYaST modules (nested arrays and hashes).
+        #
+        # This method provides no extra validation, type conversion or
+        # initialization to default values. Those responsibilities belong to the
+        # AutoYaST modules. The hash is expected to be valid and
+        # contain the relevant information. Attributes are set to nil for
+        # missing keys and for blank values.
+        #
+        # @param hash   [Hash] content of the corresponding section of the profile.
+        #   Each element of the hash corresponds to one of the attributes
+        #   defined in the section.
+        # @param parent [#parent,#section_name] parent section
+        # @return [SectionWithAttributes]
+        def new_from_hashes(hash, parent = nil)
+          result = new(parent)
+          result.init_from_hashes(hash)
+          result
+        end
+
+      protected
+
+        # Macro used in the subclasses to define accessors for all the
+        # attributes defined by {.attributes}
+        def define_attr_accessors
```

Comment:
I play a bit with `inherited` callbacks so not all class need to define it themself, but it is sadly called before class is defined, so cannot be used. Maybe we found nicer solution then to call it in all subclasses? Maybe method missing or OpenStruct?

Reply:
I am usually fine with calling the method explicitly and avoiding magic :) However, I agree that we could find a better way.

Reply:
calling method if you need something is fine for me. But having to call it always in each child looks like unnecessary syntax glue for me.